src/test/RequestQueue.test.ts

```diff
@@ -0,0 +1,357 @@
+      promises.push(rq.addRequest(loadSpec, work));
+      await Promise.all(promises);
+      expect(count).to.equal(1);
+    });
+
+    // Test that same promise is returned
+
+    it("handles identical key objects", async () => {
+      const rq = new RequestQueue();
+      const loadSpec1 = new LoadSpec();
+      const loadSpec2 = new LoadSpec();
+
+      let count = 0;
+      const promises: Promise<unknown>[] = [];
+      const work = async () => {
+        await sleep(100);
+        count++;
+      };
+      const promise1 = rq.addRequest(loadSpec1, work);
+      const promise2 = rq.addRequest(loadSpec2, work);
+      // Check that the promises are the same instance
+      expect(promise1).to.deep.equal(promise2);
+      promises.push(promise1);
+      promises.push(promise2);
+      await Promise.all(promises);
+      expect(count).to.equal(1);
+    });
+
+    it("handles multiple concurrent requests", async () => {
+      const rq = new RequestQueue();
+      const array = [false, false, false, false, false];
+      const promises: Promise<unknown>[] = [];
+      for (let i = 0; i < array.length; i++) {
+        const work = async () => {
+          await sleep(10);
+          array[i] = true;
+        };
+        promises.push(rq.addRequest(i, work));
+      }
+      await Promise.all(promises);
+      for (let i = 0; i < array.length; i++) {
+        expect(array[i]).to.be.true;
+      }
+    });
+
+    it("completes all tasks sequentially when max requests is set", async () => {
+      const rq = new RequestQueue(JSON.stringify, 1);
+      const startTime = Date.now();
+      const iterations = 5;
+      const delayMs = 5;
+      const counter: number[] = [];
+
+      const promises: Promise<unknown>[] = [];
+      for (let i = 0; i < iterations; i++) {
+        const work = async () => {
+          await sleep(delayMs);
+          counter.push(i);
+        };
+        promises.push(rq.addRequest(i, work));
+      }
+      await Promise.all(promises);
+
+      // Should run only one task at a time, so time should be
+      // at LEAST the timeout * number of tasks
+      const duration = Date.now() - startTime;
+      expect(duration).to.be.greaterThan(delayMs * iterations);
+
+      // Tasks should execute in sequential order, all tasks should run.
+      for (let i = 0; i < iterations; i++) {
+        expect(counter[i]).to.equal(i);
+      }
+    });
+
+    it("handles failing request actions", async () => {
+      const maxActiveRequests = 10;
+      const rq = new RequestQueue(JSON.stringify, maxActiveRequests);
+      const iterations = maxActiveRequests * 10;
+
+      const promises: Promise<unknown>[] = [];
+      for (let i = 0; i < iterations; i++) {
+        promises.push(
+          rq.addRequest(i, async () => {
+            await sleep(5);
+            throw new Error("Test error (should be caught)");
+          })
+        );
+      }
+      await Promise.allSettled(promises).catch((_) => {
```

Comment:
maybe slightly more interesting/nontrivial to have half fail and half succeed too.

Reply:
It would also be good to make sure that an error message passed out via reject can be received (do string comparison, as if the failures were for different reasons).

Reply:
Added! This was a useful activity for seeing how promises handle errors too :)